src/dsp_tools/utils/xmlupload/xmlupload.py

```diff
@@ -345,6 +417,153 @@ def _upload_resources(
     return id2iri_mapping, failed_uploads, metrics
 
 
+def _upload_multimedia_to_sipi(
+    resource: XMLResource,
+    sipi_server: Sipi,
+    imgdir: str,
+    filesize: float,
+    permissions_lookup: dict[str, Permissions],
+    metrics: list[MetricRecord],
+    preprocessing_done: bool,
+) -> dict[str, str | Permissions] | None:
+    """
+    This function takes a resource with a corresponding bitstream filepath.
+    If the pre-processing is not done, it retrieves the file from the directory and uploads it to sipi.
+    If pre-processing is done it retrieves the bitstream information from sipi.
+
+    Args:
+        resource: resource with that has a bitstream
+        sipi_server: server to upload
+        imgdir: directory of the file
+        filesize: size of the file
+        permissions_lookup: dictionary that contains the permission name as string and the corresponding Python object
+        metrics: to store metric information in
+        preprocessing_done: If True, then no upload is necessary
+
+    Returns:
+        The information from sipi which is needed to establish a link from the resource
+    """
+    if preprocessing_done:
+        resource_bitstream = resource.get_bitstream(
+            internal_file_name_bitstream=resource.bitstream.value,  # type: ignore[union-attr]
+            permissions_lookup=permissions_lookup,
+        )
+    else:
+        pth = resource.bitstream.value  # type: ignore[union-attr]
+        bitstream_start = datetime.now()
+        filetype = Path(pth).suffix[1:]
+        img: Optional[dict[Any, Any]] = try_network_action(
+            sipi_server.upload_bitstream,
+            filepath=str(Path(imgdir) / Path(pth)),
+        )
+        bitstream_duration = datetime.now() - bitstream_start
+        bitstream_duration_ms = bitstream_duration.seconds * 1000 + int(bitstream_duration.microseconds / 1000)
+        mb_per_sec = round((filesize / bitstream_duration_ms) * 1000, 1)
+        metrics.append(
+            MetricRecord(resource.id, filetype, filesize, "bitstream upload", bitstream_duration_ms, mb_per_sec)
+        )
+
+        internal_file_name_bitstream = img["uploadedFiles"][0]["internalFilename"]  # type: ignore[index]
+        resource_bitstream = resource.get_bitstream(
+            internal_file_name_bitstream=internal_file_name_bitstream,
+            permissions_lookup=permissions_lookup,
+        )
+    return resource_bitstream
+
+
+def _check_for_bitstream_and_upload_if_there(
```

Comment:
this function feels like a detour, also it is somewhat misleading because it says that it does an upload, which is not necessarily true (when preprocessing is done), but it can still produce an error (when getting the bitstream information), which would then be handled as an "upload error"